src/main/scala/is/hail/asm4s/Code.scala

```diff
@@ -250,11 +253,47 @@ class CodeBoolean(val lhs: Code[Boolean]) extends AnyVal {
     }
   }
 
-  def &&(rhs: Code[Boolean]): Code[Boolean] =
+  def &(rhs: Code[Boolean]): Code[Boolean] =
     Code(lhs, rhs, new InsnNode(IAND))
 
-  def ||(rhs: Code[Boolean]): Code[Boolean] =
+  def &&(rhs: Code[Boolean]): Code[Boolean] = {
+    new Code[Boolean] {
```

Comment:
I haven't seriously checked the practical benefits, but we've tried to use `CodeConditional` instead of explicitly constructing booleans so that when booleans are used in a `mux` we never have to `LDC 1`.

I think this is just:

```scala
new CodeConditional {
  def emitConditional(il: Growable[AbstractInsnNode]): (LabelNode, LabelNode) = {
    val lfalse = new LabelNode()
    val ltrue = new LabelNode()
    lhs.emit(il)
    il += new JumpInsnNode(IFEQ, lfalse)
    rhs.emit(il)
    il += new JumpInsnNode(IFEQ, lfalse)
    il += new JumpInsnNode(GOTO, ltrue)
    (ltrue, lfalse)
  }
}
```

Reply:
I think you can do slightly better:

```
class CodeConditional { self =>
  def &&(rhs: CodeConditional) = new CodeConditional {
    def emitConditional(il: ...) = {
      val (lt1, lf1) = self.emitConditional(il)
      il += lt1
      val (lt2, lf2) = rhs.emitConditional(il)
      il += lf1
      il += GotoNode(lf2)
      (lt2, lf2)
    }
  }
}
```

Reply:
So move && from CodeBoolean to CodeConditional?

Reply:
In fact, reflecting on this, `emitConditional` should take the two labels as input.  Then you can get rid of the "lf1: goto lf2".

Reply:
Both, I think.  This on `CodeConditional`, then `def &&(rhs: CodeBoolean) = toConditional && rhs.toConditional`.  `toConditional` does nothing if it is already conditional.